### DISCUSSION TITLE
Account recovery with code strategy

### DISCUSSION THREAD

**Discussion by qrest** [ory/kratos#2923] [Q&A] [answered]
Hi, 

I'm unable to get the new code strategy for account recovery working. 
I upgraded to kratos v0.11 (including DB migration) and I'm using the new SDK via Javascript in a Browser. 
I'm initializing the recovery flow as usual and render the form which accepts the code. 
On submit I'm calling  `updateRecoveryFlow({ flow: id, updateRecoveryFlowBody: body })`, where `body` is the form body. This results in a HTTP POST to `/self-service/recovery?flow=<flow_id>` with 
`` {"csrf_token":"<token>","code":" 244714","method":"code","email":"test@test.com"}`` 
as the body. 

Instead of a successful recovery, I'm getting 
``An email containing a recovery code has been sent to the email address you provided``. Note that at this point I already got an email with a code. So this seems wrong. 

I tried to remove the `email` field from the HTTP request, which results in 
``The recovery code is invalid or has already been used. Please try again``. I'm positive that the used code is correct. 

With the link strategy everything still works.

Any ideas what I'm doing wrong?
`

**Comment by jonas-jonas** [comment, marked as answer]
Hi,

it seems that what's happening here is that, the `email` field is not supposed to be sent when a code is submitted. In regular HTML based browsers, the "form" mechanism only submits either the `code` or the `email` field. Which environment are you using?

From your description, it seems that you first submitted the body with both the `email` and `code` field, which invalidates the used code, as the system assumes that you're requesting a new code. After that, you removed the `email` field from the body and tried to submit the now invalid code again, which resulted in the `The recovery code is invalid or has already been used. Please try again.` message. Is that correct?

In any case, please try to create a new flow, submit the email on the first input field, so you get the `An email containing a recovery code has been sent to the email address you provided.` message and then submit the code, but make sure the `body` payload does not contain the `email` field.

**Reply by qrest** [reply]
Hi, 
thank you for the quick response.

I found my mistake: Copying the recovery code from the email included a leading space. After trimming the input, the account recovery was successful.

**Reply by DASPRiD** [reply]
I ran into the same issue. It's weird that kratos does not trim the input, guess I gonna do that on the frontend then.
